man/man1/patchwork.1
.\"    This file is part of the FElt finite element analysis package.
.\"    Copyright (C) 1993-1996 Jason I. Gobat and Darren C. Atkinson
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.TH PATCHWORK 1fe "5/19/94" "Version 2.20" "Finite Element Package"
.SH NAME
patchwork \- file format conversion application
.SH SYNOPSIS
.B patchwork
[\-help]
[\-iformat \fIfilename\fR]
[\-oformat \fIfilename\fR]
[\-nocpp]
[\-cpp \fIfilename\fR]
.SH DESCRIPTION
\fIPatchwork\fR is a program to convert between different types of problem
description file formats. Obviously, it is mainly intended as a method to go 
back and forth between standard \fIfelt\fR(4fe) files and files formatted 
for input into other popular systems (e.g., DXF), but it could just as well 
be used to go between two of the alternative formats. In the above synopsis, 
\fBiformat\fR and \fBoformat\fR should be replaced with appropriate input and 
output format styles (see below).
.SH OPTIONS
\fIPatchwork\fR accepts the following options:
.TP 5
.B \-help
Write a message describing valid command line options to standard error
and exit.
.TP
.BI \-iformat " filename"
Currently, \fBiformat\fR should be one of \fBdxf\fR, \fBfelt\fR, or
\fBgraph\fR, specifying that the input is either a DXF file, a standard FElt 
input file, or a data file for a graphing application like \fIgnuplot\fR,
respectively.  A filename of "\-" indicates that input should be read from 
standard input.
.TP
.BI \-oformat " filename"
Currently, \fBoformat\fR should be one of \fBdxf\fR, \fBfelt\fR, \fBgraph\fR, 
or \fBlogan\fR, specifying that the resulting output should be formatted as
DXF, FElt, as input for \fIgnuplot\fR, or as appropriate input for the 
software that accompanies D.L. Logan's introductory finite element text.  A 
filename of "\-" indicates that the output should be written to standard 
output.
.TP
.B \-nocpp
Do not use a preprocessor on the input file.
.TP
.BI \-cpp " filename"
Use \fIfilename\fR as a preprocessor on the input file.  The default
preprocessor is "/lib/cpp".  Any preprocssor which understands the -D, -U,
and -I options can be used as these options are passed to the preprocessor.
.SH AUTHOR
\fIPatchwork\fR was developed by Jason I. Gobat (jgobat@mit.edu) and Darren
C. Atkinson (atkinson@ucsd.edu).
.SH SEE ALSO
felt(4fe).


CHANGELOG
** Changes to FElt from 3.02 to 3.05 (first version released on SourceForge)

The following changes were made to the FElt syntax:
   1. changed the sense of coordinates for axisymmetric elements to
      make them easier to use with mesh generation.  x is still radial
      but now y is vertical/axial.  Previously z was vertical.  

The following bugs/things were fixed/changed:
   1. some small bugs in axisymmetric, ctg, htk, and iso_2d elements
      that I haven't been really good about keeping tack of.  Try
      the new version and see if your answers look better -- if they
      look worse then let me know ...
   2. you can now use the -details switch to get the coordinates of
      the points where stresses are calculated in the text display of
      results.
   3. added radial and axial directions to the load dialog in velvet
      and to the file writing routines.  They'd gotten left out when
      axisymmetric elements were originally implemented.
   4. finally resolved the colormap issue in velvet so you can 
      use color stress and displacement plots on > 8-bit displays.
      Take a look at src/Velvet/XCC.README to see how you too can
      accomplish this if you're X11 colormap impaired like us.
   5. dropped Geompk in favor of Triangle by J.R. Shewchuk.  Several
      reasons for the change: Geompk was always a little finicky,
      mostly because of our interface probably, but it was sometimes
      difficult to coax a mesh out of Geompk.  The interface to
      Triangle is very simple, it is coded in C which makes compilation
      and configuration on other architectures simple, and it seems
      to be pretty easy to get a decent mesh out of it.  Accordingly,
      the main input parameters have changed to target and alpha.
      target is the approximate number of desired elements in the mesh.
      alpha is used to generate a maximum bound on the area of a generated
      element: Amax of an elt = alpha*Atotal/target.  Triangle appears to
      have a lot of other options for mesh refinement, angle constraints
      etc., but none of these are made use of at this point.  Most of the
      hooks for Geompk are still available if someone really wants to
      switch back.  The call to GenerateTriMesh did not change for 
      applications, but the applications did change to account for
      the loss of Geompk parameters (min, max, angspc, angtol, etc.)
      and the addition of target and alpha.
     
** Changes to FElt from 3.00 or 3.01 to 3.02 

The following changes were made to the FElt syntax:
   1. hooks for a gravity vector in the analysis section.  Doesn't
      do anything yet, but the syntax is there ...
   2. "static-substitution" and "static-incremental" are now valid 
      analysis types
   3. a "load cases" section and accompanying syntax for named
      load cases
   4. syntax for new analysis parameters, "tolerance", "iterations",
      "relaxation", "load-steps".
   5. input-node= and input-dof= parameters in the analysis 
      parameters to allow for parametric studies in which you
      get a table (or plot) of output that explores displacements
      as a function of force level at a single input point
   6. axial and radial are now valid directions for distributed loads

The following bugs/things were fixed/changed:
   1. the transform matrix for beam3d elements was incorrect if elements
      had nodes which were out of the x-y plane.
   2. several problems with the new Makefile scheme were corrected.
   3. velvet was probably getting the aspect ratios wrong in drawing 
      3d structures due to an error in the min/max checking.
   4. the spectrum components of force structures were not being
      initialized to either zero or NULL
   5. save's in velvet were actually doing dumps (save all's), minor, 
      but potentially annoying.
   6. there was still a problem with the lumped mass matrices for 
      htk elements.
   7. the new PrintGlobalMatrices routine took a file pointer but
      didn't actually use it; things were just going to stdout.
   8. internal error in burlap's interface wit the matrix lib
      was causing some problems with inversion and overwriting.
   9. there were problems with iso_2d elements (or any element with
      optional nodes) as the renumbering routines were just blindly
      looping through element -> definition -> numnodes rather than
      checking to make sure that all of the nodes were in fact valid.
  10. on a similar note, the checks in velvet for out-of-plane nodes
      were going over element -> definition -> numnodes rather than
      element -> definition -> shapenodes (which all have to be valid
      in the first place).
  11. there was some sort of problem with the pre-processor and
      corduroy (don't ask me how these things happen).  I think it's
      fixed now, but then again, I don't really know what was wrong.
  12. renamed AssembleStiffness and AssembleDynamic to ConstructStiffness
      and ConstructDynamic to reflect that they both create and assemble
      the matrices (this then parallels ConstructForceVector).  Renamed
      ConstructTransientForce to AssembleTransientForce to reflect 
      the fact it only fills the thing rather than creating it then
      filling it.
  13. the cpp command line wasn't being built properly so you couldn't
      actually use -I, -D and -U on the felt (and corduroy) command line
  14. memory initialization problem in Drawing.c
  15. 2d and 3d Athena widget specific defaults were moved into
      velvet.c and are now applied automatically at runtime as
      appropriate.
  16. the filename printing and scrolling in velvet's file dialog
      was improved.
  17. changed the analysis.nodes array into an array of actual node
      pointers rather than node numbers.  In some ways this adds
      a layer of complexity because whenever we input analysis nodes
      we have to resolve their numbers by doing tree lookups in the

src/Patchwork/patchwork.c
/*
    This file is part of the FElt finite element analysis package.
    Copyright (C) 1993-2000 Jason I. Gobat and Darren C. Atkinson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/****************************************************************************
 *
 * File:         patchwork.c
 *
 * Description:  Contains code for the driver application for the format
 *               conversion package for FElt
 *
 *****************************************************************************/

# include <stdio.h>
# include <string.h>
# include "problem.h"
# include "error.h"
# include "allocate.h"
# include "patchwork.h"
# include "definition.h"

# define streq(a,b)	!strcmp(a,b)

# define FELT  1
# define DXF   2
# define LOGAN 3
# define GRAPH 4
# define OOGL  5

# ifndef DOS
static char *usage = "\
usage: patchwork -[in format] [filename] -[out format] [filename] [options]\n\
\n\
       Available input formats:\n\
       dxf           AutoCAD style DXF files\n\
       felt          standard FElt files\n\
       graph         standard ASCII graph data (gnuplot)\n\
\n\
       Available output formats:\n\
       dxf           AutoCAD style DXF files\n\
       felt          standard FElt files\n\
       graph         standard ASCII graph data (gnuplot)\n\
       logan         files formatted for the software from Logan's book\n\
\n\
       Available pre-processor options for FElt input files:\n\
       -nocpp              do not use a preprocessor\n\
       -cpp filename       preprocessor to use\n\
       -Dname[=value]      define a macro\n\
       -Uname              undefine a macro\n\
       -Idirectory         specify include directory\n\
";
# else
static char *usage = "\
usage: patchwork -[in format] [filename] -[out format] [filename] [options]\n\
\n\
       Available input formats:\n\
       dxf           AutoCAD style DXF files\n\
       felt          standard FElt files\n\
       graph         standard ASCII graph data (gnuplot)\n\
\n\
       Available output formats:\n\
       dxf           AutoCAD style DXF files\n\
       felt          standard FElt files\n\
       graph         standard ASCII graph data (gnuplot)\n\
       logan         files formatted for the software from Logan's book\n\
       oogl          the OOGL Geom File Format to view with Geomview\n\
";
# endif

static char *input_name = NULL;
static char *output_name = NULL;

static int input_format = 0;
static int output_format = 0;

/************************************************************************
 * Function:	ParseOptions						*
 *									*
 * Description:	Parses the command line options.			*
 ************************************************************************/

static int ParseOptions (argc, argv)
    int  *argc;
    char *argv [ ];
{
    int   	i;
    int   	j;
    char 	*arg;
    int	  	format;

    j = 1;
    for (i = 1; i < *argc; i ++) {
        format = 0;
	if (streq ((arg = argv [i]), "-help")) {
	    fputs (usage, stderr);
	    exit (0);
	} else if (streq (arg, "-felt")) {
	    if (++ i == *argc) 
		return 1;

            format = FELT; 
        } else if (streq (arg, "-dxf")) {
            if (++ i == *argc) 
               return 1;

            format = DXF; 
        } else if (streq (arg, "-logan")) {
            if (++ i == *argc)
               return 1;
            
            format = LOGAN; 
        } else if (streq (arg, "-graph")) {
            if (++ i == *argc)
               return 1;
            
            format = GRAPH; 
        } else if (streq (arg, "-oogl")) {
            if (++ i == *argc)
               return 1;
            
            format = OOGL; 
	} else
	    argv [j ++] = arg;

        if (format) {
           if (input_format) {
              output_format = format;
              output_name = argv [i];
           }
           else {
              input_format = format;
              input_name = argv [i];
           } 
        }
    }

    argv [*argc = j] = NULL;
    return 0;
}


/************************************************************************
 * Function:	 main							*
 *									*
 * Description:	 Main is the driver function for the patchwork 		*
 *		 application.						*
 ************************************************************************/

int main (argc, argv)
    int   argc;
    char *argv [ ];
{
    int	  status;

	/*
	 * what the heck - just in case the input file is FElt style
	 */

# ifndef DOS
    if (ParseCppOptions (&argc, argv)) {
        fputs (usage, stderr);
        exit (1);
    }
# endif

	/*
	 * get the patchwork specific options
	 */

    if (ParseOptions (&argc, argv)) {
	fputs (usage, stderr);
	exit (1);
    }

    if (!input_format || !output_format || !input_name || !output_name) {
	fputs (usage, stderr);
	exit (1);
    }

    add_all_definitions ( );

	/*
	 * read the input file
	 */

    if (input_format == FELT)
       status = ReadFeltFile (input_name);
    else if (input_format == DXF)
       status = ReadDXFFile (input_name);
    else if (input_format == GRAPH)
       status = ReadGraphFile (input_name);
    else {
       error ("unrecognized input format.");
       status = 1;
    }
 
    if (status)
       exit (1);

	/*
	 * write the ouput file
	 */

    if (output_format == FELT)
       DumpFeltFile (output_name);
    else if (output_format == DXF) 
       WriteDXFFile (output_name);
    else if (output_format == GRAPH) 
       WriteGraphFile (output_name);
    else if (output_format == LOGAN) 
       WriteLoganFile (output_name);
    else if (output_format == OOGL) 
       WriteOoglFile (output_name);
    else {
       error ("unrecognized output format.");
       exit (1);
    }

    exit (0);
}
